### ISSUE TITLE
Aggressive Results Using Many `--include` Flags

### ISSUE THREAD

**Issue by shellscape** [duniul/clean-modules#4]
In preparing a fill-in for #3, I've converted our original `.yarnclean` into something that `picomatch` can use. I'm parsing that config, removing comments, and building a list of `-i` flags which represent globs from our config file. It appears that `clean-modules` sees the flags correctly.

When I don't use these flags, `clean-modules` correctly trims about 80mb from `node_modules`. When I use these flags, _everything in node_modules is removed_. I've run the list of globs through `picomatch` directly and it correctly identified the files it should based on those globs. `picomatch` did not match the same number of files that `clean-modules` does. Something is amiss.

Here's the analysis output:

```
Analyzed results:
{
  globs: {
    defaultDirs: '**/@(__tests__|test|tests|powered-test|docs|doc|.idea|.vscode|example|examples|coverage|.nyc_output|.nycrc|.circleci|.github|.git|website)',
    defaultFiles: '**/@(*.test.*|*.map|*.@(md|mkd|markdown|mdown)|*.mustache|*.asciidoc|*.DOCS|!(*.d).ts|*.coffee|*.zip|*.7z|*.rar|*.tar|*.tgz|*.@(jpg|jpeg)|*.png|*.gif|*.h|*.c|*.hpp|*.cpp|*.o|*.mk|*.log|*.tlog|*.patch|*.sln|*.pdb|*.jst|*.swp|*.lock|*.vcxproj*|makefile*|gemfile*|gulpfile*|gruntfile*|rakefile*|Jenkinsfile|.travis.yml|.gitlab-ci.yml|.appveyor.yml|circle.yml|.DS_Store|.eslintrc*|.stylelintrc*|stylelint.config.js|.htmllintrc*|htmllint.js|.jshintrc*|.lint|tslint.json|.prettierrc*|prettier.config.js|test.js|jest.config.js|karma.conf.js|wallaby.js|wallaby.conf.js|.coveralls.yml|readme*|changelog*|changes*|authors*|contributors*|contributing*|notice*|license|licence|tsconfig.json|.tsbuildinfo|.npmrc|.*ignore|yarn.lock|.yarnclean|.yarn-metadata.json|.gitmodules|.gitattributes|.babelrc|.editorconfig|.tern-project|.flowconfig|.documentup.json|.yo-rc.json|_config.yml|bower.json|.vimrc*|binding.gyp|component.json|composer.json|.jamignore|.jscsrc|*.todo|*.md|*.markdown|contributors|*.orig|*.rej|.zuul.yml|.editorconfig|.npmrc|.jshintignore|.eslintignore|.lint|.lintignore|cakefile|.istanbul.yml)',
    includeArgs: [
      '**/__tests__',
      '**/tests',
      '**/powered-test',
      '**/*.test.js',
      '**/*.tap.js',
      '**/*.test.ts',
      '**/*tests.html',
      '**/docs',
      '**/doc',
      '**/website',
      '**/images',
      '**/assets',
      '!**/@aws-cdk/**/assets',
      '!**/@aws-cdk/**/images',
      '**/example',
      '**/examples',
      '**/coverage',
      '**/.nyc_output',
      '**/Makefile',
      '**/Gulpfile.js',
      '**/Gruntfile.js',
      '**/.github/**/*.yml',
      '**/.*ignore',
      '**/.babelrc',
      '**/.browserslistrc',
      '**/.dist.babelrc',
      '**/.dist.eslintrc',
      '**/.documentup.json',
      '**/.editorconfig',
      '**/.eslintrc',
      '**/.eslintrc.js',
      '**/.eslintrc.json',
      '**/.eslintrc.yml',
      '**/.flowconfig',
      '**/.gitattributes',
      '**/.idea',
      '**/.jshintrc',
      '**/.lib.babelrc',
      '**/.lib.eslintrc',
      '**/.tern-project',
      '**/.travis.yml',
      '**/.yarn-metadata.json',
      '**/.zuul.yml',
      '**/appveyor.yml',
      '**/bower.json',
      '**/circle.yml',
      '**/codeship-services.yml',
      '**/codeship-steps.yml',
      '**/jsdoc_conf.json',
      '**/karma-ci.conf.js',
      '**/rollup.config.js',
      '**/tsconfig.json',
      '**/webpack.config.js',
      '**/wercker.yml',
      '**/*.html',
      '**/*.gif',
      '**/*.jpg',
      '**/*.lock',
      '**/*.log',
      '**/*.markdown',
      '**/*.map',
      '**/*.md',
      '**/*.png',
      '**/*.sh',
      '**/*.sln',
      '**/*.txt',
      '**/*.un~',
      '**/AUTHORS',
      '**/LICENSE',
      '**/MIT-LICENSE'
    ],
    excludeArgs: undefined
  },
```

It's worth nothing that if I build the `--include` flag mirroring the `defaultFiles` property, I get the correct number of excluded files:

```
Analyzed results:
{
  globs: {
    defaultDirs: '**/@(__tests__|test|tests|powered-test|docs|doc|.idea|.vscode|example|examples|coverage|.nyc_output|.nycrc|.circleci|.github|.git|website)',
    defaultFiles: '**/@(*.test.*|*.map|*.@(md|mkd|markdown|mdown)|*.mustache|*.asciidoc|*.DOCS|!(*.d).ts|*.coffee|*.zip|*.7z|*.rar|*.tar|*.tgz|*.@(jpg|jpeg)|*.png|*.gif|*.h|*.c|*.hpp|*.cpp|*.o|*.mk|*.log|*.tlog|*.patch|*.sln|*.pdb|*.jst|*.swp|*.lock|*.vcxproj*|makefile*|gemfile*|gulpfile*|gruntfile*|rakefile*|Jenkinsfile|.travis.yml|.gitlab-ci.yml|.appveyor.yml|circle.yml|.DS_Store|.eslintrc*|.stylelintrc*|stylelint.config.js|.htmllintrc*|htmllint.js|.jshintrc*|.lint|tslint.json|.prettierrc*|prettier.config.js|test.js|jest.config.js|karma.conf.js|wallaby.js|wallaby.conf.js|.coveralls.yml|readme*|changelog*|changes*|authors*|contributors*|contributing*|notice*|license|licence|tsconfig.json|.tsbuildinfo|.npmrc|.*ignore|yarn.lock|.yarnclean|.yarn-metadata.json|.gitmodules|.gitattributes|.babelrc|.editorconfig|.tern-project|.flowconfig|.documentup.json|.yo-rc.json|_config.yml|bower.json|.vimrc*|binding.gyp|component.json|composer.json|.jamignore|.jscsrc|*.todo|*.md|*.markdown|contributors|*.orig|*.rej|.zuul.yml|.editorconfig|.npmrc|.jshintignore|.eslintignore|.lint|.lintignore|cakefile|.istanbul.yml)',
    includeArgs: [
      '**/@(,__tests__|tests|powered-test|*.test.js|*.tap.js|*.test.ts|*tests.html|docs|doc|website|images|assets|!**/@aws-cdk/**/assets|!**/@aws-cdk/**/images|example|examples|coverage|.nyc_output|Makefile|Gulpfile.js|Gruntfile.js|.github/**/*.yml|.*ignore|.babelrc|.browserslistrc|.dist.babelrc|.dist.eslintrc|.documentup.json|.editorconfig|.eslintrc|.eslintrc.js|.eslintrc.json|.eslintrc.yml|.flowconfig|.gitattributes|.idea|.jshintrc|.lib.babelrc|.lib.eslintrc|.tern-project|.travis.yml|.yarn-metadata.json|.zuul.yml|appveyor.yml|bower.json|circle.yml|codeship-services.yml|codeship-steps.yml|jsdoc_conf.json|karma-ci.conf.js|rollup.config.js|tsconfig.json|webpack.config.js|wercker.yml|*.html|*.gif|*.jpg|*.lock|*.log|*.markdown|*.map|*.md|*.png|*.sh|*.sln|*.txt|*.un~|AUTHORS|LICENSE|MIT-LICENSE,)'
    ],
    excludeArgs: undefined
  },
```

**Comment by duniul** [comment]
@shellscape

**TL/DR**: [Use the `--exclude` flag](https://github.com/duniul/clean-modules#--exclude---e) instead of prefixing inclusions with `!` 👍 

This is an example of the `picomatch` vs. `gitignore`/`yarnclean` syntax mismatch that I mentioned in the other comment.

It's happening because you are trying to exclude files by prefixing inclusion globs with `!`, but that actually gets interpreted by `picomatch` as "include any path that does not match this glob". For example, when passing `!**/@aws-cdk/**/assets` it will include everything except a nested `assets` directory in the `@aws-cdk` scope!

This is the expected behavior of `picomatch`, here's an example:

```js
const picomatch = require('picomatch');

const isMatch = picomatch([
  'only/include/this',
]);

const isMatchWithExcludeGlobs = picomatch([
  'only/include/this',
  '!**/@aws-cdk/**/assets',
  '!**/@aws-cdk/**/images',
]);

console.log(isMatch('only/include/this')); // true
console.log(isMatch('not/this')); // false
console.log(isMatch('or/this')); // false

console.log(isMatchWithExcludeGlobs('only/include/this')); // true
console.log(isMatchWithExcludeGlobs('not/this')); // true
console.log(isMatchWithExcludeGlobs('or/this')); // true
```

You can get around this by passing your exclusion globs with the [`--exclude` or `-e` flag](https://github.com/duniul/clean-modules#--exclude---e) (or `ignore` to `picomatch`). With the included args from your example, I guess it would be something like:

```
yarn clean-modules \ 
  -i "**/__tests__" -i "**/tests" -i "**/powered-test" -i "**/*.test.js" -i "**/*.tap.js" -i "**/*.test.ts" -i "**/*tests.html" -i "**/docs" -i "**/doc" -i "**/website" -i "**/images" -i "**/assets" -i "**/example" -i "**/examples" -i "**/coverage" -i "**/.nyc_output" -i "**/Makefile" -i "**/Gulpfile.js" -i "**/Gruntfile.js" -i "**/.github/**/*.yml" -i "**/.*ignore" -i "**/.babelrc" -i "**/.browserslistrc" -i "**/.dist.babelrc" -i "**/.dist.eslintrc" -i "**/.documentup.json" -i "**/.editorconfig" -i "**/.eslintrc" -i "**/.eslintrc.js" -i "**/.eslintrc.json" -i "**/.eslintrc.yml" -i "**/.flowconfig" -i "**/.gitattributes" -i "**/.idea" -i "**/.jshintrc" -i "**/.lib.babelrc" -i "**/.lib.eslintrc" -i "**/.tern-project" -i "**/.travis.yml" -i "**/.yarn-metadata.json" -i "**/.zuul.yml" -i "**/appveyor.yml" -i "**/bower.json" -i "**/circle.yml" -i "**/codeship-services.yml" -i "**/codeship-steps.yml" -i "**/jsdoc_conf.json" -i "**/karma-ci.conf.js" -i "**/rollup.config.js" -i "**/tsconfig.json" -i "**/webpack.config.js" -i "**/wercker.yml" -i "**/*.html" -i "**/*.gif" -i "**/*.jpg" -i "**/*.lock" -i "**/*.log" -i "**/*.markdown" -i "**/*.map" -i "**/*.md" -i "**/*.png" -i "**/*.sh" -i "**/*.sln" -i "**/*.txt" -i "**/*.un~" -i "**/AUTHORS" -i "**/LICENSE" -i "**/MIT-LICENSE" \
  -e "**/@aws-cdk/**/assets" -e "**/@aws-cdk/**/images"
```

**BUT!** With that amount of custom inclusions I'd definitely recommend wrapping them into one glob with a single globstar, which drastically speeds up `picomatch`. In that case you'd run this instead:

```
yarn clean-modules \
  -i "**/@(__tests__|tests|powered-test|*.test.js|*.tap.js|*.test.ts|*tests.html|docs|doc|website|images|assets|example|examples|coverage|.nyc_output|Makefile|Gulpfile.js|Gruntfile.js|.github/**/*.yml|.*ignore|.babelrc|.browserslistrc|.dist.babelrc|.dist.eslintrc|.documentup.json|.editorconfig|.eslintrc|.eslintrc.js|.eslintrc.json|.eslintrc.yml|.flowconfig|.gitattributes|.idea|.jshintrc|.lib.babelrc|.lib.eslintrc|.tern-project|.travis.yml|.yarn-metadata.json|.zuul.yml|appveyor.yml|bower.json|circle.yml|codeship-services.yml|codeship-steps.yml|jsdoc_conf.json|karma-ci.conf.js|rollup.config.js|tsconfig.json|webpack.config.js|wercker.yml|*.html|*.gif|*.jpg|*.lock|*.log|*.markdown|*.map|*.md|*.png|*.sh|*.sln|*.txt|*.un~|AUTHORS|LICENSE|MIT-LICENSE)" \
  -e "**/@aws-cdk/**/@(assets|images)"
```
I tried comparing this command with the one above on a monorepo with about 1 GB of `node_modules` and it went from ~5 seconds to ~1.4. Every second counts in CI 🔥 

In the next version (which hopefully also includes #3) I hope to do this automatically for matching patterns, although I don't want to tamper with the user's custom patterns too much.

**Comment by shellscape** [comment]
Yeah I ended up wrapping them, going off of the `DEFAULT_FILES` in code, and was able to keep the negations in there, just as you have the `!*.d.ts` entry in `DEFAULT_FILES`. Worked a charm there.
